flocker/node/agents/blockdevice.py

```diff
@@ -1705,7 +1786,23 @@ def discover_state(self, cluster_state, persistent_state):
         datasets = {}
         for volume in raw_state.volumes:
             dataset_id = volume.dataset_id
-            if dataset_id in raw_state.devices:
+            owning_blockdevice_id = persistent_state.blockdevice_ownership.get(
+                dataset_id)
+            if owning_blockdevice_id is None:
+                datasets[dataset_id] = DiscoveredDataset(
+                    state=DatasetStates.UNREGISTGERED,
+                    dataset_id=dataset_id,
+                    maximum_size=volume.size,
+                    blockdevice_id=volume.blockdevice_id,
+                )
+            elif volume.blockdevice_id != owning_blockdevice_id:
+                # XXX Should we cleanup duplicate volumes?
```

Comment:
I'm going to leave the comment for now. I think dealing with existing duplicate volumes is out of the scope of this issue (and definitely out of scope of this PR). I'll expand the comment to explain that if we get to that point, then it *is* a duplicate volume.